tests/tests_unit/test_http_client.py

```diff
@@ -0,0 +1,155 @@
+
+import pytest
+import requests.exceptions
+import urllib3.exceptions
+
+from cognite.client._http_client import HTTPClient, HTTPClientConfig, _RetryTracker
+from cognite.client.exceptions import CogniteConnectionError, CogniteConnectionRefused, CogniteReadTimeout
+
+DEFAULT_CONFIG = HTTPClientConfig(
+    status_codes_to_retry={429},
+    backoff_factor=0.5,
+    max_backoff_seconds=30,
+    max_retries_total=10,
+    max_retries_read=5,
+    max_retries_connect=5,
+    max_retries_status=10,
+)
+
+
+class TestRetryTracker:
+    def test_total_retries_exceeded(self):
+        rt = _RetryTracker(config=DEFAULT_CONFIG)
+        rt.config.max_retries_total = 10
+        rt.status = 4
+        rt.connect = 4
+        rt.read = 4
+
+        assert rt.total == 12
+        assert rt.should_retry(200) is False
+
+    def test_status_retries_exceeded(self):
+        rt = _RetryTracker(config=DEFAULT_CONFIG)
+        rt.config.max_retries_status = 1
+        assert rt.should_retry(None) is True
+        rt.status = 1
+        assert rt.should_retry(None) is False
+
+    def test_read_retries_exceeded(self):
+        rt = _RetryTracker(config=DEFAULT_CONFIG)
+        rt.config.max_retries_read = 1
+        assert rt.should_retry(None) is True
+        rt.read = 1
+        assert rt.should_retry(None) is False
+
+    def test_connect_retries_exceeded(self):
+        rt = _RetryTracker(config=DEFAULT_CONFIG)
+        rt.config.max_retries_connect = 1
+        assert rt.should_retry(None) is True
+        rt.connect = 1
+        assert rt.should_retry(None) is False
+
+    def test_correct_retry_of_status(self):
+        rt = _RetryTracker(config=DEFAULT_CONFIG)
+        assert rt.should_retry(500) is False
+        rt.config.status_codes_to_retry = {500, 429}
+        assert rt.should_retry(500) is True
+
+    def test_get_backoff_time(self):
+        rt = _RetryTracker(config=DEFAULT_CONFIG)
+        for i in range(1000):
+            rt.read += 1
+            assert 0 <= rt.get_backoff_time() <= DEFAULT_CONFIG.max_backoff_seconds
+
+
+def raise_exception_wrapped_as_in_requests_lib(exc: Exception):
+    try:
+        raise exc
+    except type(exc):
+        try:
+            raise urllib3.exceptions.RequestError(pool=None, url=None, message=None)
+        except urllib3.exceptions.RequestError:
+            try:
+                raise urllib3.exceptions.MaxRetryError(pool=None, url=None)
+            except urllib3.exceptions.MaxRetryError:
+                raise requests.exceptions.ReadTimeout
+
+
+class TestHTTPClient:
+    def test_read_timeout_errors(self):
+        cnf = DEFAULT_CONFIG
+        cnf.max_backoff_seconds = 0
+        c = HTTPClient(
+            config=cnf,
+            session=MagicMock(
+                request=MagicMock(
+                    side_effect=lambda *args, **kwargs: raise_exception_wrapped_as_in_requests_lib(socket.timeout())
+                )
+            ),
+        )
+
+        with pytest.raises(CogniteReadTimeout):
```

Comment:
this strikes me as really odd; either use unittest tests and support libraries, or pytest testing idioms, mixing both is liable to confuse well, everyone :). Not least test suite tooling authors like me :P.

Reply:
I thought we were mostly using pytest with unittest mainly for mock, the entire test suite is like this.

Reply:
Yeah, not sure what is odd here. Only using MagicMock from unittest, otherwise this is all pytest.

Reply:
(This class is not a unittest.TestCase)